File Block: Copy url button is moved to Block toolbar

daddou-ma opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->
closes #26328
## Description
<!-- Please describe what you have changed or added -->
As mentioned in issue #26328 the **copy url** button in the **file block** is editor only element. so in the PR i moved the **copy url button** to the **block toolbar**.

## How has this been tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
### Before
![image](https://user-images.githubusercontent.com/4710635/96619013-ef91b500-1305-11eb-9721-23968bd4a0ff.png)
### After
![image](https://user-images.githubusercontent.com/15234847/97716313-bf93a000-1ac3-11eb-8be5-4ecb23998c96.png)

## Types of changes
### Enhancement
 - File Block: Copy url button is moved to Block toolbar
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->

daddou-ma committed:
File Block: Copy url button is moved to Block toolbar

daddou-ma committed:
Experiments local demo test fix: wait for the selector before testing it

adding the `page.waitForSelector()` function to wait for the `button.editor-post-save-draft` selector to load before testing it

daddou-ma commented:
@senadir @youknowriad i have the following problem in e2e tests. can you help me ? 

After adding the daddou-ma/gutenberg@3e68718 commit i got the `packages/e2e-tests/specs/local/demo.test.js` (E2E Admin 3) test failing even if my changes are not concerned by this test. 

So i've consulted the test and fixed it by adding `page.waitForSelector` function before test daddou-ma/gutenberg@13979ba. but after this second commit i got others tests failling in (E2E Admin 1)
```
packages/e2e-tests/specs/experiments/template-part.test.js
packages/e2e-tests/specs/editor/plugins/allowed-blocks.test.js
packages/e2e-tests/specs/local/demo.test.js
```

daddou-ma commented:
> @senadir @youknowriad i have the following problem in e2e tests. can you help me ?
> 
> After adding the [daddou-ma/gutenberg@3e68718](https://github.com/daddou-ma/gutenberg/commit/3e68718) commit i got the `packages/e2e-tests/specs/local/demo.test.js` (E2E Admin 3) test failing even if my changes are not concerned by this test.
> 
> So i've consulted the test and fixed it by adding `page.waitForSelector` function before test [daddou-ma/gutenberg@13979ba](https://github.com/daddou-ma/gutenberg/commit/13979ba). but after this second commit i got others tests failling in (E2E Admin 1)
> 
> ```
> packages/e2e-tests/specs/experiments/template-part.test.js
> packages/e2e-tests/specs/editor/plugins/allowed-blocks.test.js
> packages/e2e-tests/specs/local/demo.test.js
> ```

same issue is happening in Wordpress/gutenberg#26605 and Wordpress/gutenberg#26611 PRs

youknowriad commented on packages/block-library/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### Enhancement
+
+- File Block: Copy url button is moved to Block toolbar
```
```suggestion
- File Block: Copy url button is moved to Block toolbar.
```

youknowriad reviewed (commented):
The code change is looking good but I noticed that when using arrow navigation in the toolbar, that particular button is skipped for some reason?

Also curious what designers think about this. I agree it's better than what we have in master, that button shouldn't be in the content but it does increase the size of the toolbar a bit.

jasmussen commented:
Thank you for the PR, thank you for the screenshots! 

I would agree with Riad, it seems a decently appropriate place to put the button given it's a primary action. I also noticed that the Bold/Italic buttons do not appear until you actually set focus inside the Download button, so most of the time the toolbar isn't quite as wide:

<img width="1103" alt="Screenshot 2020-11-02 at 10 43 53" src="https://user-images.githubusercontent.com/1204802/97853462-6fe9ea00-1cf8-11eb-8847-21b9ed92c45b.png">

I looked for a "copy URL" icon, but did not find a sufficiently good one. It does seem like a text button, as suggested here, may be the best approach. It does make the toolbar wider, but I'm softly leaning towards "let's try it". Curious if others have thoughts.

ellatrix commented on packages/block-library/src/file/edit.js:
```diff
@@ -171,13 +173,28 @@ function FileEdit( {
+						onError={ onUploadError }
+					/>
+					<ToolbarItem
+						as={ ClipboardButton }
```
Maybe we could try the new `useCopyOnClick` hook here?

daddou-ma replied:
I tried to use the `useCopyOnClick` hook, but it will trigger an error because it requires using the toolbar button `ref` which is not be rendered yet when the file component loads, then `ref` value will be undefined. 

In case of toolbar which is rendered only on focus `useCopyOnClick` hook requires using a container component which is the exact role of `ClipboardButton`.

ellatrix replied:
What if you make it a separate component so `useCopyOnClick` is only called when the button is rendered? `ClipboardButton` just uses `useCopyOnClick` internally, so it should be possible to swap it out.

daddou-ma replied:
Thank you @ellatrix  for your help. I got your point, i agree it's better to use `useCopyOnClick` hook and reduce code inside 'FileEdit' Component. i've switched to `useCopyOnClick` in daddou-ma/gutenberg@f3d98d7 commit.

ellatrix commented:
This is much better than having the button in the content. It also gets rid the layout shifting at certain breakpoints when selecting and deselecting the block.

daddou-ma commented:
@youknowriad, @jasmussen, @ellatrix  Thanks for your feedback. this encourage me to work more in this project.

> The code change is looking good but I noticed that when using arrow navigation in the toolbar, that particular button is skipped for some reason?
>
>Also curious what designers think about this. I agree it's better than what we have in master, that button shouldn't be in the content but it does increase the size of the toolbar a bit.

Yes @youknowriad  i checking why it's skipping this particular button. i think it's because of `ClipboardButton`.

> Thank you for the PR, thank you for the screenshots!
>
> I would agree with Riad, it seems a decently appropriate place to put the button given it's a primary action. I also noticed that the Bold/Italic buttons do not appear until you actually set focus inside the Download button, so most of the time the toolbar isn't quite as wide:
> 
> <img alt="Screenshot 2020-11-02 at 10 43 53" width="1103" src="https://user-images.githubusercontent.com/1204802/97853462-6fe9ea00-1cf8-11eb-8847-21b9ed92c45b.png">
> 
> I looked for a "copy URL" icon, but did not find a sufficiently good one. It does seem like a text button, as suggested here, may be the best approach. It does make the toolbar wider, but I'm softly leaning towards "let's try it". Curious if others have thoughts.

You're Welcome @jasmussen. Thank you for checking my PR. for me i found `dashicons-paperclip` and `dashicons-clipboard` icons but i don't think they are the rights ones. 
![image](https://user-images.githubusercontent.com/15234847/97955195-ee509580-1da5-11eb-8e46-47462c456078.png)


> This is much better than having the button in the content. It also gets rid the layout shifting at certain breakpoints when selecting and deselecting the block.

I've just remembered that i have to check if there is CSS associated with he old Copy URL button.

daddou-ma committed:
File Copy URL Button: remove old css.

senadir commented:
> I found dashicons-paperclip and dashicons-clipboard icons but i don't think they are the rights ones.
Gutenberg is not using dashicons anymore, so those are not a good fit.

> i checking why it's skipping this particular button. i think it's because of ClipboardButton.
I think that unless we fix this, it would be a regression.

daddou-ma commented:
> Gutenberg is not using dashicons anymore, so those are not a good fit.

Which icons can i use @senadir @jasmussen ? text buttons are taking lot of space. and part of the toolbar is hidden in mobile screen.

![image](https://user-images.githubusercontent.com/15234847/97973541-83fc1d00-1dc6-11eb-9192-014c7dbb792b.png)
 

> I think that unless we fix this, it would be a regression.

i've already fixed this i will commit the changes.

jasmussen commented:
Thank you for thinking of mobile.

I would not use any icons, because there just isn't a good one. If anything, we should reduce the space between these two buttons:

<img width="250" alt="Screenshot 2020-11-03 at 11 24 22" src="https://user-images.githubusercontent.com/1204802/97973944-2a482280-1dc7-11eb-9bd2-ed72ab335269.png">

But that should be done separately, as it needs to be a general fix for all such text buttons.

On mobile, you can swipe the toolbar horizontally, so it's not the end of the world.

daddou-ma committed:
File Block: fix arrow navigation skipping of Copy URL Button.

Copy URL Button was skipped in arrow navigation because it was inserted as Button instead of ToolbarButton, so to fix this i made the following change
ClipBoardButton: added `as: Component` props so the Button Component can be overridden, so it can be inserted as ToolbarButton.

daddou-ma committed:
Update packages/block-library/CHANGELOG.md

Co-authored-by: Riad Benguella <benguella@gmail.com>

daddou-ma commented:
> The code change is looking good but I noticed that when using arrow navigation in the toolbar, that particular button is skipped for some reason?

@youknowriad, i've fixed Arrow navigation in daddou-ma/gutenberg@531ddb8. `ClipboardButton` should be used as `ToobarButton`.

daddou-ma committed:
File Block: use useCopyOnClick hook instead of ClipboardButton

ellatrix commented on packages/block-library/src/file/edit.js:
```diff
@@ -30,13 +32,21 @@ import { file as icon } from '@wordpress/icons';
-	noticeUI,
-	noticeOperations,
-} ) {
+const ClipboardToolbarButton = function ClipboardToolbarButton( { text } ) {
```
```suggestion
function ClipboardToolbarButton( { text } ) {
```

ellatrix commented on packages/components/src/clipboard-button/index.js:
```diff
@@ -15,6 +15,7 @@ import { useCopyOnClick } from '@wordpress/compose';
 import Button from '../button';
 
 export default function ClipboardButton( {
+	as: Component = Button,
```
We can now remove this change, right? :)

daddou-ma replied:
Yes we don't need it. but need help !, am i allowed to remove the commit containing this change?  daddou-ma/gutenberg@1be6587 or i should make another commit to remove it ?

daddou-ma replied:
removed in daddou-ma/gutenberg@ebb7fee

ellatrix commented on packages/block-library/src/file/edit.js:
```diff
@@ -171,13 +168,19 @@ function FileEdit( {
+					/>
+					<ClipboardToolbarButton
+						text={ href }
+						disabled={ isBlobURL( href ) }
```
I don't see this prop being used in the component

daddou-ma replied:
i added it as prop in daddou-ma/gutenberg@ebb7fee

ellatrix reviewed (approved):
Looks like we just need to wait for e2e tests to stabilise. Can't get it to pass even after some reruns.

daddou-ma commented:
> Looks like we just need to wait for e2e tests to stabilise. Can't get it to pass even after some reruns.

@ellatrix @youknowriad @jasmussen @senadir Thanks to all of you, I appreciate your help. I've learned a lot from this experience ^^.
